Add: make modal windows update more smooth and fix "closing the app" issues closely related to this.

TrueBrain opened the pull request:
- This invalidates the first three commits in #8828, but not the fourth. That is, this also fixes the bug mentioned, but not in the way you think: it simply continues the NewGRF scan till it is finished, while the window is closed.
- EXTRA EXTRA EXTRA: the second commit in this PR also fixes #8760, and stops the NewGRF scan as soon as possible, similar to #8828 (but without all the complex thread-stuff, as it is now running in the game-thread anyway).
- And while at it, fixed similar issue for world-gen: that is now also aborted when you close the game.
- #8833 was closely related to all this, and now also part of this PR

(okay, this might be growing a bit too much; if it is easier if I PR them separately (a few depend on each other, but I can deal with that), just let me know!)

Closes #8828
Fixes #8760
Fixes #8833

## Motivation / Problem

With the gameloop now threaded, we can get ride of the ugliness that the NewGRFScan and GenerateWorld threads are. They are a huge hack dating back when threads were still scary and not many people had more than one core. Pretty sure those days are over :P

By removing the special code for those two modal windows, the code becomes a lot simpler. Having such massive chunks of code not running in different threads makes things easier to debug and understand.

Purely by accident, this solves most issues mentioned in #8712, with the exception of the palette.

But mostly ... it now looks so much more smooth to generate a world ...

https://user-images.githubusercontent.com/1663690/110473423-cb7d1780-80de-11eb-947c-d22d4344ea0c.mp4


## Description

```
Basically, modal windows had their own thread-locking for what
drawing was possible. This is a bit nonsense now we have a
game-thread. And it makes much more sense to do things like
NewGRFScan and GenerateWorld in the game-thread, and not in a
thread next to the game-thread.

This commit changes that: it removes the threads for NewGRFScan
and GenerateWorld, and just runs the code in the game-thread.
On regular intervals it allows the draw-thread to do a tick,
which gives a much smoother look and feel.

It does slow down NewGRFScan and GenerateWorld ever so slightly
as it spends more time on drawing. But the slowdown is not
measureable on my machines (with 700+ NewGRFs / 4kx4k map and
a Debug build).

Running without a game-thread means NewGRFScan and GenerateWorld
are now blocking.
```

All the extra `MarkDirty` calls are because formally it just made the whole screen dirty every time, but that is very expensive. So now only that what changed is marked dirty, which is a lot nicer on the CPU.

## Limitations

- Slightly slower NewGRF scanning and map generation, because time is spend on drawing (and not on scanning). On my machine it is so small, I could not measure this outside the noise. I would estimate it at: 0.05ms per frame, 60 frames per second, 3ms per 1000ms, so 0.3% speed loss.
- Running with `-vwin32-opengl:no_threads` means you don't see those two modals; similar to running without threads before this commit. I think this is reasonable, to make it work this way.
- The game used to mark the whole screen dirty during map generation; but doing that now would be dreadfully slow. So instead I tried to find all the places that modify the map, and mark the files dirty. To me it seems I found them all, but I might have missed some. Nothing we can't fix over time. (Basically, the old implementation was lazy and just did EVERYTHING, instead of what is needed).


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

TrueBrain committed:
Add: make modal windows update more smooth

Basically, modal windows had their own thread-locking for what
drawing was possible. This is a bit nonsense now we have a
game-thread. And it makes much more sense to do things like
NewGRFScan and GenerateWorld in the game-thread, and not in a
thread next to the game-thread.

This commit changes that: it removes the threads for NewGRFScan
and GenerateWorld, and just runs the code in the game-thread.
On regular intervals it allows the draw-thread to do a tick,
which gives a much smoother look and feel.

It does slow down NewGRFScan and GenerateWorld ever so slightly
as it spends more time on drawing. But the slowdown is not
measureable on my machines (with 700+ NewGRFs / 4kx4k map and
a Debug build).

Running without a game-thread means NewGRFScan and GenerateWorld
are now blocking.

TrueBrain committed:
Fix: abort GRFFileScanner on exiting the game as soon as possible

This prevents the window from "freezing" when you close it during
the scanning of NewGRFs, as it first would continue the action.

TrueBrain committed:
Fix: abort world generation on exiting the game as soon as possible

This prevents the window from "freezing" when you close it during
world generation, as it first would continue the action.

ldpl commented:
Some stuff I noticed after testing (some may not be related to this pr but whatever):

- Doesn't crash if closed during newgrf scan
- Reaches 100% way before finishing the scan. Few times I've seen it starting with 100% and doing entire scan on there.
- Landscape generation part of mapgen still lags as before
- Sleeping cloud cursor is still quite confusing to use (when clicking abort button).
- Mapgen seems to count as world tick for frame rate window so it initializes avg with some insanely high number.

TrueBrain commented:
Tnx for testing!

>     * Doesn't crash if closed during newgrf scan
> 

Good! That is the important bit :D

>     * Reaches 100% way before finishing the scan. Few times I've seen it starting with 100% and doing entire scan on there.
> 

Are you sure it is doing the scan at 100%? How do you tell?

For me it is working as expected, but one important observation: at the end of the NewGRF Scan, it stalls for a bit of time. There is a pretty hefty function being called when the Scan is complete, which stalls for anywhere between 300ms up to 2seconds for me.

>     * Landscape generation part of mapgen still lags as before
> 

The first 5 steps are indeed laggy. This is because TGP does a massive calculation at once, and prints that on the map. There are no in-betweens there, and it doesn't allow mouse movement while doing. I think future-extensions can be to allow mouse movements, even during times game-tick has a lock on the game-state. But we are a bit away from that :)

>     * Sleeping cloud cursor is still quite confusing to use (when clicking abort button).
> 

*shrug*. Not disagreeing, but indeed out of scope of this PR. Maybe worth an issue on its own?

>     * Mapgen seems to count as world tick for fps so it initializes avg with some insanely high number.
>

Haha, I noticed that too :D I will fix this one, as it is really silly :P

TrueBrain committed:
Fix #8833: don't reload NewGRFs when we are shutting down

Otherwise that might cause calls to the video-driver, which are
already shut down by now. This causes, depending on the video-driver
crashes or weird effects.

TrueBrain commented:
>     * Reaches 100% way before finishing the scan. Few times I've seen it starting with 100% and doing entire scan on there.
> 

Basically, the game remembers how many NewGRFs you had last round, and assumes you will have that again. So the 100% is at the last known amount you had (`last_newgrf_count` in the config). My PR used to store the value when you closed the window, but in the next push it will no longer do that, and only remember the number of a finished scan.

TrueBrain committed:
Fix: don't update framerates when a modal window is open

Otherwise the numbers are all over the place when a modal window
just closed.

LordAro commented on src/openttd.h:
```diff
@@ -81,4 +81,13 @@ void HandleExitGameRequest();
+ * done in the game-thread, and not in the draw-thread (which most often
+ * triggers this request).
+ * @param callback Optional callback to call when NewGRF scan is completed.
+ */
```
Shouldn't this go on the definition, rather than the declaration?

LordAro commented on src/video/video_driver.cpp:
```diff
@@ -56,6 +56,20 @@ void VideoDriver::GameThread()
+	}
+
+	this->game_state_mutex.lock();
+}
```
unlocking, then locking & immediately unlocking, then locking again without doing anything else inbetween looks really  weird :)

TrueBrain replied:
That is what you get if OSes decide to optimize mutexes, and don't give other locks priority when you unlock :P Really stupid. Added a comment about it.

LordAro commented on src/video/video_driver.hpp:
```diff
@@ -179,6 +179,12 @@ class VideoDriver : public Driver {
 
+	/**
+	 * Pause the game-loop for a bit, releasing the game-state locks, so
+	 * drawing can happen.
```
weird linebreak

LordAro reviewed (commented):
I like the look of the code in general, all sorts of mutex & global removals. As for it's correctness...

LordAro reviewed (approved):
It's been tested reasonably thoroughly, and removing mutexes & global variables is always nice

YOLO
